<a href="https://colab.research.google.com/github/ruchithareddy269/Dm-Assignment-3/blob/main/SEMMA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install pycaret

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 483.9/483.9 kB 7.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 42.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.4/153.4 kB 8.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.8/81.8 kB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 19.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 12.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.8/106.8 kB 13.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.4/73.4 kB 8.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 77.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 4.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 81.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 68.6 MB/s eta 0:

In [7]:
#mounting google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Please find the attached dataset-https://www.kaggle.com/datasets/janiobachmann/bank-marketing-dataset


In [10]:
# Load the dataset
bank_data= pd.read_csv('/content/drive/MyDrive/DM-Assignment-2-part2/bank.csv')

In [11]:
bank_data.head()

age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes

The dataset seems to contain information about bank customers and whether or not they have made a deposit. The attributes include details like age, job, marital status, education, balance, and other related factors.

Based on the data provided, a potential goal for our project could be to predict whether a customer will make a deposit based on their details.


Given this goal, let's walk through the SEMMA methodology:

# **1. Sample**

Although we typically sample data when the dataset is very large, for the sake of this demonstration, let's assume we're working with the full dataset.


# **2. Explore**

We'll start by exploring the dataset further to understand its distributions, identify any missing values, and recognize any potential outliers or anomalies.

In [12]:
# Checking basic statistics and information about the dataset
summary_info = bank_data.describe(include='all')

# Checking for missing values in the dataset
missing_values = bank_data.isnull().sum()

summary_info, missing_values

(                 age         job  marital  education default       balance  \
 count   11162.000000       11162    11162      11162   11162  11162.000000   
 unique           NaN          12        3          4       2           NaN   
 top              NaN  management  married  secondary      no           NaN   
 freq             NaN        2566     6351       5476   10994           NaN   
 mean       41.231948         NaN      NaN        NaN     NaN   1528.538524   
 std        11.913369         NaN      NaN        NaN     NaN   3225.413326   
 min        18.000000         NaN      NaN        NaN     NaN  -6847.000000   
 25%        32.000000         NaN      NaN        NaN     NaN    122.000000   
 50%        39.000000         NaN      NaN        NaN     NaN    550.000000   
 75%        49.000000         NaN      NaN        NaN     NaN   1708.000000   
 max        95.000000         NaN      NaN        NaN     NaN  81204.000000   
 
        housing   loan   contact           day  mo

From our exploration:

The dataset has 11,162 records.
There are no missing values in the dataset.
Variables such as age, balance, duration, campaign, pdays, and previous are numerical, and we've obtained statistics like mean, median, and standard deviation for them.
Other variables are categorical, including the target variable deposit.

# **3. Modify**

Based on our exploration, we can identify modifications that might be needed:

Categorical variables like job, marital, education, default, housing, loan, contact, month, poutcome, and deposit may need to be encoded.
Some numerical variables might benefit from normalization or scaling.
We can also derive new features if necessary.
Let's start by encoding the categorical variables and scaling the numerical ones.

In [13]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Encoding categorical variables
label_encoders = {}
categorical_columns = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome', 'deposit']

for col in categorical_columns:
    le = LabelEncoder()
    bank_data[col] = le.fit_transform(bank_data[col])
    label_encoders[col] = le

# Scaling numerical variables
scaler = StandardScaler()
numerical_columns = ['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']
bank_data[numerical_columns] = scaler.fit_transform(bank_data[numerical_columns])

# Displaying the modified dataset
bank_data.head()

age  job  marital  education  default   balance  housing  loan  \
0  1.491505    0        1          1        0  0.252525        1     0   
1  1.239676    0        1          1        0 -0.459974        0     0   
2 -0.019470    9        1          1        0 -0.080160        1     0   
3  1.155733    7        1          1        0  0.293762        1     0   
4  1.071790    0        1          2        0 -0.416876        0     0   

   contact       day  month  duration  campaign     pdays  previous  poutcome  \
0        2 -1.265746      8  1.930226 -0.554168 -0.481184  -0.36326         3   
1        2 -1.265746      8  3.154612 -0.554168 -0.481184  -0.36326         3   
2        2 -1.265746      8  2.929901 -0.554168 -0.481184  -0.36326         3   
3        2 -1.265746      8  0.596366 -0.554168 -0.481184  -0.36326         3   
4        2 -1.265746      8  0.867171 -0.186785 -0.481184  -0.36326         3   

   deposit  
0        1  
1        1  
2        1  
3        1  
4        1

# **4. Model**

For modeling, we'll:

Split the data into training and testing sets.
Build a simple predictive model using a decision tree classifier as a starting point.
Train the model using the training set.
Let's start by splitting the data.

In [14]:
from sklearn.model_selection import train_test_split

# Splitting the data into training and testing sets (80% training, 20% testing)
X = bank_data.drop('deposit', axis=1)
y = bank_data['deposit']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape

((8929, 16), (2233, 16))

The data has been split into training (8,929 samples) and testing (2,233 samples) sets.

Now, let's build and train a Decision Tree classifier using the training data.

In [15]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Build and train a Decision Tree classifier
dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train, y_train)

# Predict on the test set
y_pred = dt_classifier.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.7639946260635916

# **5. Assess**

To assess our model further, let's evaluate it using additional metrics like precision, recall, and the F1-score. These metrics will give us a more comprehensive understanding of the model's performance, especially if our dataset has class imbalances.

In [16]:
from sklearn.metrics import classification_report

# Generate classification report for the Decision Tree classifier
report = classification_report(y_test, y_pred, target_names=['No Deposit', 'Deposit'])
report

'              precision    recall  f1-score   support\n\n  No Deposit       0.77      0.79      0.78      1166\n     Deposit       0.76      0.74      0.75      1067\n\n    accuracy                           0.76      2233\n   macro avg       0.76      0.76      0.76      2233\nweighted avg       0.76      0.76      0.76      2233\n'

The model's performance, based on the classification report, is as follows:

Precision:
No Deposit:
0.77
0.77

Deposit:
0.76
0.76

Recall:
No Deposit:
0.79
0.79
Deposit:
0.74
0.74


F1-Score:
No Deposit:
0.78
0.78
Deposit:
0.75
0.75
Overall Accuracy:
0.76
0.76
This provides a more detailed view of how our model performs for each class (Deposit vs. No Deposit).